### PR TITLE
CRUD functional implementation

### LINKED ISSUE #3: Implement CRUD functional for todo micro-service

### PR DESCRIPTION
**Pull request by MaksimPegov**

### CODE REVIEW

File: todos/src/main/java/com/maksimpegov/todos/TodosService.java [hunk 1]
```diff
@@ -1,35 +1,95 @@
...
+        try {
+            if (!todoRepository.existsById(Long.parseLong(todoId))) {
+                response.setStatus("404");
+                return response;
+            }
+            response.setTodos(Collections.singletonList(todoRepository.getOne(Long.parseLong(todoId))));
+            todoRepository.deleteById(Long.parseLong(todoId));
+            response.setStatus("200");
+        } catch (Exception e) {
+            System.out.println(e.getMessage());
```
Comment (antonpegov) [inline todos/src/main/java/com/maksimpegov/todos/TodosService.java:68]: Is this is only for real-time debugging? Maybe it is make sence to add logging to the project?

File: todos/src/main/java/com/maksimpegov/todos/TodosService.java [hunk 2]
```diff
@@ -1,35 +1,95 @@
...
+        }
+        return response;
     }
 
-    public String deleteTodo(String userId, String todoId) {
-        todoRepository.deleteById(Long.parseLong(todoId));
-        return "OK";
+    public ProcessResponse editTodo(Todo newTodo) {
+        ProcessResponse response = new ProcessResponse();
+        try {
```
Comment (antonpegov) [inline todos/src/main/java/com/maksimpegov/todos/TodosService.java:37]: Just a proposual for better readability. You can use empty lines to divide logical parts. I woud use it before 'try', then befor 'if' and before return...

File: todos/src/main/java/com/maksimpegov/todos/TodosService.java [hunk 3]
```diff
@@ -1,35 +1,95 @@
...
+            Todo oldTodo = todoRepository.getOne(newTodo.getId());
+            oldTodo.setCreatedAt(new Date());
```
Comment (antonpegov) [inline todos/src/main/java/com/maksimpegov/todos/TodosService.java]: Maybe it is make sence to add 'lastEditAt' field and use it here?

File: todos/src/main/java/com/maksimpegov/todos/TodosService.java [hunk 4]
```diff
@@ -1,35 +1,95 @@
...
+            if (newTodo.getText() != null) {
+                oldTodo.setText(newTodo.getText());
+            } else if (newTodo.getClosedAt() != null) {
```
Comment (antonpegov) [inline todos/src/main/java/com/maksimpegov/todos/TodosService.java]: this whole part with if else if is looks strange... It is needed only to merge two objects of class ToDo, rignt?
If yes, then the code is really looks bad. There is no way to understend from the code that newTodo is only partial ToDo. You need another way to merge newTodo and oldTodo objects.

File: todos/src/main/java/com/maksimpegov/todos/TodosService.java [hunk 5]
```diff
@@ -1,35 +1,95 @@
...
+                oldTodo.setClosedAt(newTodo.getClosedAt());
+            } else if (newTodo.isCompleted()) {
+                oldTodo.setCompleted(true);
+                oldTodo.setClosedAt(new Date());
+            } else {
+                response.setStatus("400");
```
Comment (antonpegov) [inline todos/src/main/java/com/maksimpegov/todos/TodosService.java]: Pls give the user some message so he can understand that he send wrong request.

File: todos/src/main/java/com/maksimpegov/todos/models/ProcessResponse.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package com.maksimpegov.todos.models;
+
+import java.util.Collections;
+import java.util.List;
+
+public class ProcessResponse {
```
Comment (antonpegov) [inline todos/src/main/java/com/maksimpegov/todos/models/ProcessResponse.java]: What means 'ProcessResponse'? Mybe better use 'TodoServiceResponse'? Just curious... Oh you have TodoService responce already.... So why then TodoService returning ProcessResponse? And controller returning TodoServiceResponse? )))

**Review by antonpegov — CHANGES_REQUESTED** [review]

**Review by antonpegov — CHANGES_REQUESTED** [review]
Some changes are preferable...

File: todos/src/main/java/com/maksimpegov/todos/TodosService.java (hunk 2 above)
Comment (MaksimPegov) [reply todos/src/main/java/com/maksimpegov/todos/TodosService.java:37]: ok

File: todos/src/main/java/com/maksimpegov/todos/TodosService.java (hunk 3 above)
Comment (MaksimPegov) [reply todos/src/main/java/com/maksimpegov/todos/TodosService.java]: alright

File: todos/src/main/java/com/maksimpegov/todos/TodosService.java (hunk 4 above)
Comment (MaksimPegov) [reply todos/src/main/java/com/maksimpegov/todos/TodosService.java]: Yes, this is the only way. The reason is, that i need to check witch field user want to edit, and replace only this part.

File: todos/src/main/java/com/maksimpegov/todos/models/ProcessResponse.java (hunk 6 above)
Comment (MaksimPegov) [reply todos/src/main/java/com/maksimpegov/todos/models/ProcessResponse.java]: TodoServiceResponse is form that whole micro-service return, ProcessResponse - is response that service gives to controller after all data work and processed

**Review by antonpegov — APPROVED** [review]

**Pull request merged**